File: ui-ngx/src/app/core/http/api-key.service.ts
///
/// Copyright © 2016-2026 The Thingsboard Authors
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.
///

import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { defaultHttpOptionsFromConfig, RequestConfig } from '@core/http/http-utils';
import { Observable } from 'rxjs';
import { PageLink } from '@shared/models/page/page-link';
import { PageData } from '@shared/models/page/page-data';
import { ApiKeyInfo, ApiKey } from '@shared/models/api-key.models';

/**
 * Angular injectable service: api key (HTTP service layer).
 *
 * <p>HTTP wrappers in `@core/http` calling ThingsBoard REST API.
 */
@Injectable({
  providedIn: 'root'
})
export class ApiKeyService {

  constructor(
    private http: HttpClient
  ) {
  }

  
  /**
   * POST/PUT entity — save api key.
   *
   * @param apiKey api key (ApiKeyInfo)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<ApiKey> observable or value
   */


  public saveApiKey(apiKey: ApiKeyInfo, config?: RequestConfig): Observable<ApiKey> {
    return this.http.post<ApiKey>('/api/apiKey', apiKey, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * DELETE — delete api key.
   *
   * @param id id (string)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<void> observable or value
   */


  public deleteApiKey(id: string, config?: RequestConfig): Observable<void> {
    return this.http.delete<void>(`/api/apiKey/${id}`, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * update api key description.
   *
   * @param id id (string)
   * @param description description (string)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<ApiKeyInfo> observable or value
   */


  public updateApiKeyDescription(id: string, description: string, config?: RequestConfig): Observable<ApiKeyInfo> {
    return this.http.put<ApiKeyInfo>(`/api/apiKey/${id}/description`, description, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * enable api key.
   *
   * @param id id (string)
   * @param enabledValue enabled value (boolean)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<ApiKeyInfo> observable or value
   */


  public enableApiKey(id: string, enabledValue: boolean, config?: RequestConfig): Observable<ApiKeyInfo> {
    return this.http.put<ApiKeyInfo>(`/api/apiKey/${id}/enabled/${enabledValue}`, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get user api keys.
   *
   * @param userId user UUID
   * @param pageLink pagination and sort parameters
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<PageData<ApiKeyInfo>> observable or value
   */


  public getUserApiKeys(userId: string, pageLink: PageLink, config?: RequestConfig): Observable<PageData<ApiKeyInfo>> {
    return this.http.get<PageData<ApiKeyInfo>>(`/api/apiKeys/${userId}${pageLink.toQuery()}`, defaultHttpOptionsFromConfig(config));
  }
}

File: common/dao-api/src/main/java/org/thingsboard/server/dao/pat/ApiKeyService.java
/**
 * Copyright © 2016-2026 The Thingsboard Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.thingsboard.server.dao.pat;

import org.thingsboard.server.common.data.id.ApiKeyId;
import org.thingsboard.server.common.data.id.TenantId;
import org.thingsboard.server.common.data.id.UserId;
import org.thingsboard.server.common.data.page.PageData;
import org.thingsboard.server.common.data.page.PageLink;
import org.thingsboard.server.common.data.pat.ApiKey;
import org.thingsboard.server.common.data.pat.ApiKeyInfo;
import org.thingsboard.server.dao.entity.EntityDaoService;

import java.util.List;

/**
 * Service API for api key persistence and domain operations.
 */
public interface ApiKeyService extends EntityDaoService {

    /**
     * Saves or persists api key.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param apiKey api key ({@link ApiKeyInfo})
     * @return {@link ApiKey}
     */
    ApiKey saveApiKey(TenantId tenantId, ApiKeyInfo apiKey);

    /**
     * Saves or persists api key.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param apiKeyInfo api key info ({@link ApiKeyInfo})
     * @param value value ({@link String})
     * @param doValidate whether to run validation before persist
     * @return {@link ApiKey}
     */
    ApiKey saveApiKey(TenantId tenantId, ApiKeyInfo apiKeyInfo, String value, boolean doValidate);

    /**
     * Deletes api key.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param apiKey api key ({@link ApiKey})
     * @param force force
     */
    void deleteApiKey(TenantId tenantId, ApiKey apiKey, boolean force);

    /**
     * Deletes by user id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param userId user id ({@link UserId})
     */
    void deleteByUserId(TenantId tenantId, UserId userId);

    /**
     * Finds api key by value.
     *
     * @param value value ({@link String})
     * @return {@link ApiKey}
     */
    ApiKey findApiKeyByValue(String value);

    /**
     * Finds api key by id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param apiKeyId api key id ({@link ApiKeyId})
     * @return {@link ApiKey}
     */
    ApiKey findApiKeyById(TenantId tenantId, ApiKeyId apiKeyId);

    /**
     * Finds api keys by user id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param userId user id ({@link UserId})
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<ApiKeyInfo> findApiKeysByUserId(TenantId tenantId, UserId userId, PageLink pageLink);

    /**
     * Finds api keys by user id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param userId user id ({@link UserId})
     * @return {@link List}
     */
    List<ApiKey> findApiKeysByUserId(TenantId tenantId, UserId userId);

    /**
     * Finds api keys by tenant id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<ApiKey> findApiKeysByTenantId(TenantId tenantId, PageLink pageLink);

}
